src/vstest.console/Internal/MSBuildLogger.cs

```diff
@@ -0,0 +1,166 @@
+
+    /// <summary>
+    /// Constructor added for testing purpose
+    /// </summary>
+    internal MSBuildLogger(IOutput output)
+    {
+        Output = output;
+    }
+
+    [MemberNotNull(nameof(Output))]
+    public void Initialize(TestLoggerEvents events, string testRunDirectory)
+    {
+        ValidateArg.NotNull(events, nameof(events));
+
+        Output ??= ConsoleOutput.Instance;
+
+        // Register for the events.
+        // events.TestRunMessage += TestMessageHandler;
+        events.TestResult += TestResultHandler;
+        events.TestRunComplete += TestRunCompleteHandler;
+        // events.TestRunStart += TestRunStartHandler;
+
+        // Register for the discovery events.
+        // events.DiscoveryMessage += TestMessageHandler;
+    }
+
+    public void Initialize(TestLoggerEvents events, Dictionary<string, string?> parameters)
+    {
+        ValidateArg.NotNull(parameters, nameof(parameters));
+
+        if (parameters.Count == 0)
+        {
+            throw new ArgumentException("No default parameters added", nameof(parameters));
+        }
+
+        var verbosityExists = parameters.TryGetValue(VerbosityParam, out var verbosity);
+        if (verbosityExists && Enum.TryParse(verbosity, true, out Verbosity verbosityLevel))
+        {
+            // We don't use this anywhere for now.
+            VerbosityLevel = verbosityLevel;
+        }
+
+        Initialize(events, string.Empty);
+    }
+
+    private void TestRunCompleteHandler(object? sender, TestRunCompleteEventArgs e)
+    {
+        TPDebug.Assert(Output != null, "Initialize should have been called");
+
+        if (e.IsCanceled)
+        {
+            Output.Error(false, CommandLineResources.TestRunCanceled);
+        }
+        else if (e.IsAborted)
+        {
+            if (e.Error == null)
+            {
+                Output.Error(false, CommandLineResources.TestRunAborted);
+            }
+            else
+            {
+                Output.Error(false, CommandLineResources.TestRunAbortedWithError, e.Error);
+            }
+        }
+    }
+
+    private void TestResultHandler(object? sender, TestResultEventArgs e)
+    {
+        ValidateArg.NotNull(sender, nameof(sender));
+        ValidateArg.NotNull(e, nameof(e));
+        TPDebug.Assert(Output != null, "Initialize should have been called.");
+
+        switch (e.Result.Outcome)
+        {
+            case TestOutcome.Failed:
+                {
+                    var stringBuilder = new StringBuilder();
+                    var testDisplayName = e.Result.DisplayName;
+
+                    if (e.Result.DisplayName.IsNullOrWhiteSpace())
+                    {
+                        testDisplayName = e.Result.TestCase.DisplayName;
+                    }
+
+                    stringBuilder.Append(testDisplayName ?? "<unknown>").Append(": ");
+                    AppendFullError(e.Result, stringBuilder);
+                    Output.Error(false, stringBuilder.ToString());
```

Comment:
we could maybe emit Json or XML here, then in the MSBuild task read the whole output and parse it as Json, and emit individual errors based on the structure.